test/integration/qos_test.go

```diff
@@ -171,3 +171,112 @@ func TestQOSPlugin(t *testing.T) {
+	testCtx.Ctx, testCtx.CancelFn = context.WithCancel(context.Background())
+
+	cs := kubernetes.NewForConfigOrDie(globalKubeConfig)
+	testCtx.ClientSet = cs
+	testCtx.KubeConfig = globalKubeConfig
+
+	cfg, err := util.NewDefaultSchedulerComponentConfig()
+	if err != nil {
+		t.Fatal(err)
+	}
+	cfg.Profiles[0].Plugins.QueueSort = schedapi.PluginSet{
+		Enabled:  []schedapi.Plugin{{Name: qos.Name}},
+		Disabled: []schedapi.Plugin{{Name: "*"}},
+	}
+
+	testCtx = initTestSchedulerWithOptions(
+		t,
+		testCtx,
+		scheduler.WithProfiles(cfg.Profiles...),
+		scheduler.WithFrameworkOutOfTreeRegistry(fwkruntime.Registry{qos.Name: qos.New}),
+	)
+	syncInformerFactory(testCtx)
+	defer cleanupTest(t, testCtx)
+
+	ns := fmt.Sprintf("integration-test-%v", string(uuid.NewUUID()))
+	createNamespace(t, testCtx, ns)
+
+	// Create a Node.
+	nodeName := "fake-node"
+	node := st.MakeNode().Name(nodeName).Label("node", nodeName).Obj()
+	node.Status.Capacity = v1.ResourceList{
+		v1.ResourcePods:   *resource.NewQuantity(32, resource.DecimalSI),
+		v1.ResourceCPU:    *resource.NewMilliQuantity(500, resource.DecimalSI),
+		v1.ResourceMemory: *resource.NewQuantity(500, resource.DecimalSI),
+	}
+	node, err = cs.CoreV1().Nodes().Create(testCtx.Ctx, node, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create Node %q: %v", nodeName, err)
+	}
+
+	// Create 3 Pods with the same QoS class (Guaranteed) but different creation times.
+	var pods []*v1.Pod
+	podNames := []string{"guaranteed-1", "guaranteed-2", "guaranteed-3"}
+	pause := imageutils.GetPauseImageName()
+	for i := 0; i < len(podNames); i++ {
+		pod := st.MakePod().Namespace(ns).Name(podNames[i]).Container(pause).Obj()
+		pod.Spec.Containers[0].Resources = v1.ResourceRequirements{
+			Requests: v1.ResourceList{
+				v1.ResourceCPU:    *resource.NewMilliQuantity(100, resource.DecimalSI),
+				v1.ResourceMemory: *resource.NewQuantity(100, resource.DecimalSI),
+			},
+			Limits: v1.ResourceList{
+				v1.ResourceCPU:    *resource.NewMilliQuantity(100, resource.DecimalSI),
+				v1.ResourceMemory: *resource.NewQuantity(100, resource.DecimalSI),
+			},
+		}
+		pods = append(pods, pod)
+	}
+
+	// Create Pods sequentially with a delay between each creation to ensure different creation timestamps.
+	t.Logf("Start to create 3 Guaranteed Pods sequentially.")
+	for i, pod := range pods {
+		// Delay between creations to ensure different timestamps.
+		time.Sleep(time.Second * time.Duration(i))
+		_, err = cs.CoreV1().Pods(ns).Create(testCtx.Ctx, pod, metav1.CreateOptions{})
+		if err != nil {
+			t.Errorf("Failed to create Pod %q: %v", pod.Name, err)
+		} else {
+			t.Logf("Created Pod %q", pod.Name)
+		}
+	}
+	defer cleanupPods(t, testCtx, pods)
+
+	// Wait for all Pods are in the scheduling queue.
+	err = wait.PollUntilContextTimeout(testCtx.Ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+		pendingPods, _ := testCtx.Scheduler.SchedulingQueue.PendingPods()
```

Comment:
question: what guarantees we will find the pods in the pending pods at time of checking?
IOW, can this ever be racy?

Reply:
ok, makes sense. I need to refresh my memory of what I did in the noderesourcetopology tests but IIRC I called `Run` and here you don't, and this should explain everything